### ISSUE TITLE
[Linux/arm64] Assert failures during PMI System.Private.CoreLib

### ISSUE THREAD

**Issue by echesakovMSFT** [dotnet/coreclr#20260]
```
PREPALL type# 474 method# 8112 System.Numerics.Vector`1[[System.Numerics.Vector`1[[System.Single, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]::ToString

Assert failure(PID 14239 [0x0000379f], Thread: 14239 [0x379f]): Assertion failed 'type == lvaTable[lnum].lvType || (lvaIsImplicitByRefLocal(lnum) && fgGlobalMorph && (lvaTable[lnum].lvType == TYP_BYREF))' in 'System.Numerics.Vector`1[Vector`1][System.Numerics.Vector`1[System.Single]]:ToString(ref,ref):ref:this' (IL size 146)

    File: /opt/code/src/jit/gentree.cpp Line: 6252
    Image: /home/robox/echesako/git/_/bin/tests/Linux.arm64.Checked/Tests/Core_Root/corerun
```
The assert above fixed by #20285

```
PREPALL type# 1743 method# 22166 System.Runtime.Intrinsics.Arm.Arm64.Base::LeadingSignCount

Assert failure(PID 14246 [0x000037a6], Thread: 14246 [0x37a6]): Assertion failed '!"Unimplemented SIMD Intrinsic form"' in 'System.Runtime.Intrinsics.Arm.Arm64.Base:LeadingSignCount(int):int' (IL size 7)

    File: /opt/code/src/jit/hwintrinsicArm64.cpp Line: 324
    Image: /home/robox/echesako/git/_/bin/tests/Linux.arm64.Checked/Tests/Core_Root/corerun
```

```
PREPALL type# 1743 method# 22168 System.Runtime.Intrinsics.Arm.Arm64.Base::LeadingZeroCount

Assert failure(PID 14254 [0x000037ae], Thread: 14254 [0x37ae]): Assertion failed '!"Unimplemented SIMD Intrinsic form"' in 'System.Runtime.Intrinsics.Arm.Arm64.Base:LeadingZeroCount(int):int' (IL size 7)

    File: /opt/code/src/jit/hwintrinsicArm64.cpp Line: 324
    Image: /home/robox/echesako/git/_/bin/tests/Linux.arm64.Checked/Tests/Core_Root/corerun
```
The two assertions above fixed by #20306
```
PREPALL type# 1746 method# 22324 System.Runtime.Intrinsics.Arm.Arm64.Simd::Extract

Assert failure(PID 14268 [0x000037bc], Thread: 14268 [0x37bc]): Compiler optimization assumption invalid: ERROR: failed to find exception throw blockFAILED: (add != nullptr)
    File: /opt/code/src/jit/codegencommon.cpp Line: 2005
    Image:
```

**Steps to reproduce:**
1. Set these environement variables
```
COMPlus_TieredCompilation=0
COMPlus_ZapDisable=1
COMPlus_ReadyToRun=0
```
2. Run PMI DriveAll (needs https://github.com/dotnet/jitutils)
`$CORE_ROOT/corerun pmi.dll DRIVEALL $CORE_ROOT/System.Private.CoreLib.dll`

**Comment by echesakovMSFT** [comment]
The simplest program that would case the first assertion on Linux/arm64 is
```c#
using System;
using System.Globalization;
using System.Numerics;

namespace GitHub_20260
{
    class Program
    {
        static void Main(string[] args)
        {         
            Vector<double> x = new Vector<double>();
            string s = ((IFormattable)x).ToString("G", CultureInfo.CurrentCulture);
        }
    }
}
```

```
Assert failure(PID 41732 [0x0000a304], Thread: 41732 [0xa304]): Assertion failed 'type == lvaTable[lnum].lvType || (lvaIsImplicitByRefLocal(lnum) && fgGlobalMorph && (lvaTable[lnum].lvType == TYP_BYREF))' in 'GitHub_20260.Program:Main(ref)' (IL size 31)

    File: /opt/code/src/jit/gentree.cpp Line: 6252
    Image: 
Process 41732 stopped
* thread #1, name = 'corerun', stop reason = signal SIGTRAP
    frame #0: 0x0000ffffb38b02d0 libclrjit.so`DBG_DebugBreak
libclrjit.so`DBG_DebugBreak:
->  0xffffb38b02d0 <+0>: brk    #0
    0xffffb38b02d4 <+4>: ret

libclrjit.so`DBG_DebugBreak_End:
    0xffffb38b02d8 <+0>: stp    x29, x30, [sp, #-0x10]!
    0xffffb38b02dc <+4>: ldp    x19, x20, [x0, #0xa0]
(lldb) bt
* thread #1, name = 'corerun', stop reason = signal SIGTRAP
  * frame #0: 0x0000ffffb38b02d0 libclrjit.so`DBG_DebugBreak
    frame #1: 0x0000ffffb383aa1c libclrjit.so`::DebugBreak() at debug.cpp:405
    frame #2: 0x0000ffffb3621b48 libclrjit.so`::assertAbort(why=0x0000ffffb38e8f1d, file=0x0000ffffb38e8537, line=6252) at error.cpp:289
    frame #3: 0x0000ffffb3666b30 libclrjit.so`Compiler::gtNewLclvNode(this=0x00000000004f3b88, lnum=3, type=TYP_STRUCT, ILoffs=2147483648) at gentree.cpp:6251
    frame #4: 0x0000ffffb367a9e4 libclrjit.so`Compiler::gtTryRemoveBoxUpstreamEffects(this=0x00000000004f3b88, op=<unavailable>, options=<unavailable>) at gentree.cpp:12885
    frame #5: 0x0000ffffb36a3c28 libclrjit.so`Compiler::impDevirtualizeCall(this=<unavailable>, call=<unavailable>, method=<unavailable>, methodFlags=<unavailable>, contextHandle=<unavailable>, exactContextHandle=<unavailable>) at importer.cpp:19739
    frame #6: 0x0000ffffb36a183c libclrjit.so`Compiler::impImportCall(this=0x00000000004f3b88, opcode=CEE_CALLVIRT, pResolvedToken=0x0000ffffffffc0c0, pConstrainedResolvedToken=<unavailable>, newobjThis=<unavailable>, prefixFlags=<unavailable>, callInfo=0x0000ffffffffc1a0, rawILOffset=<unavailable>) at importer.cpp:7922
    frame #7: 0x0000ffffb36b59b0 libclrjit.so`Compiler::impImportBlockCode(this=<unavailable>, block=<unavailable>) at importer.cpp:13502
    frame #8: 0x0000ffffb36bc26c libclrjit.so`Compiler::impImportBlock(BasicBlock*) [inlined] Compiler::impImportBlock(BasicBlock*)::$_1::operator()(Compiler::impImportBlock(BasicBlock*)::FilterVerificationExceptionsParam*) const at importer.cpp:16503
    frame #9: 0x0000ffffb36bc248 libclrjit.so`Compiler::impImportBlock(this=0x00000000004f3b88, block=0x00000000004f55b0) at importer.cpp:16513
    frame #10: 0x0000ffffb36becbc libclrjit.so`Compiler::impImport(this=<unavailable>, method=<unavailable>) at importer.cpp:17588
    frame #11: 0x0000ffffb3630d64 libclrjit.so`Compiler::fgImport(this=0x00000000004f3b88) at flowgraph.cpp:6752
    frame #12: 0x0000ffffb3605224 libclrjit.so`Compiler::compCompile(this=0x00000000004f3b88, methodCodePtr=0x0000ffffffffcdc0, methodCodeSize=0x0000ffffffffd078, compileFlags=0x0000ffffffffcdd0) at compiler.cpp:4494
    frame #13: 0x0000ffffb3607a1c libclrjit.so`Compiler::compCompileHelper(this=0x00000000004f3b88, classPtr=<unavailable>, compHnd=<unavailable>, methodInfo=<unavailable>, methodCodePtr=0x0000ffffffffcdc0, methodCodeSize=0x0000ffffffffd078, compileFlags=0x0000ffffffffcdd0, instVerInfo=<unavailable>) at compiler.cpp:6082
    frame #14: 0x0000ffffb3606b2c libclrjit.so`Compiler::compCompile(CORINFO_METHOD_STRUCT_*, CORINFO_MODULE_STRUCT_*, ICorJitInfo*, CORINFO_METHOD_INFO*, void**, unsigned int*, JitFlags*) [inlined] Compiler::compCompile(CORINFO_METHOD_STRUCT_*, CORINFO_MODULE_STRUCT_*, ICorJitInfo*, CORINFO_METHOD_INFO*, void**, unsigned int*, JitFlags*)::$_1::operator()(Compiler::compCompile(CORINFO_METHOD_STRUCT_*, CORINFO_MODULE_STRUCT_*, ICorJitInfo*, CORINFO_METHOD_INFO*, void**, unsigned int*, JitFlags*)::__JITParam*) const at compiler.cpp:5440
    frame #15: 0x0000ffffb3606b08 libclrjit.so`Compiler::compCompile(this=<unavailable>, methodHnd=<unavailable>, classPtr=<unavailable>, compHnd=<unavailable>, methodInfo=<unavailable>, methodCodePtr=<unavailable>, methodCodeSize=<unavailable>, compileFlags=<unavailable>) at compiler.cpp:5460
    frame #16: 0x0000ffffb360878c libclrjit.so`jitNativeCode(CORINFO_METHOD_STRUCT_*, CORINFO_MODULE_STRUCT_*, ICorJitInfo*, CORINFO_METHOD_INFO*, void**, unsigned int*, JitFlags*, void*) at compiler.cpp:6723
    frame #17: 0x0000ffffb36086e0 libclrjit.so`jitNativeCode(CORINFO_METHOD_STRUCT_*, CORINFO_MODULE_STRUCT_*, ICorJitInfo*, CORINFO_METHOD_INFO*, void**, unsigned int*, JitFlags*, void*) at compiler.cpp:6748
(lldb) frame select 3
frame #3: 0x0000ffffb3666b30 libclrjit.so`Compiler::gtNewLclvNode(this=0x00000000004f3b88, lnum=3, type=TYP_STRUCT, ILoffs=2147483648) at gentree.cpp:6251
   6248         // Make an exception for implicit by-ref parameters during global morph, since
   6249         // their lvType has been updated to byref but their appearances have not yet all
   6250         // been rewritten and so may have struct type still.
-> 6251         assert(type == lvaTable[lnum].lvType ||
   6252                (lvaIsImplicitByRefLocal(lnum) && fgGlobalMorph && (lvaTable[lnum].lvType == TYP_BYREF)));
   6253     }
   6254     GenTree* node = new (this, GT_LCL_VAR) GenTreeLclVar(type, lnum, ILoffs);
(lldb) p type
(var_types) $5 = TYP_STRUCT
(lldb) p lnum
(unsigned int) $6 = 3
(lldb) p *(lvaTable + 3)
(LclVarDsc) $7 = {
  lvType = TYP_SIMD16
(lldb) p fgGlobalMorph
(bool) $8 = false
```

**Comment by AndyAyersMS** [comment]
I'll take a look at this first one since it is in the undo box logic I added.

**Comment by erozenfeld** [comment]
@dotnet/jit-contrib

**Comment by BruceForstall** [comment]
Before closing this, it would be useful to run PMI across all the framework assemblies to ensure no remaining asserts.

**Comment by AndyAyersMS** [comment]
Hmm, basic stuff still not happy when run on arm64. jit-diffs does not run, but prepall seems to work, mostly...
```
 /home/andya/repos/coreclr/bin/tests/Linux.arm64.Checked/Tests/Core_Root/corerun /home/andya/repos/jitutils/bin/pmi.dll PREPALL-QUIET /home/andya/repos/coreclr/bin/tests/
Linux.arm64.Checked/Tests/Core_Root/System.Private.CoreLib.dll

BadImageFormatException System.ReadOnlySpan`1[[System.Object, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]::get_Item

BadImageFormatException System.ReadOnlySpan`1[[System.Object, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]::get_Item

BadImageFormatException System.ReadOnlySpan`1[[System.Object, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]::GetPinnableReference

BadImageFormatException System.ReadOnlySpan`1+Enumerator[[System.Object, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]::get_Current

Assert failure(PID 35614 [0x00008b1e], Thread: 35614 [0x8b1e]): Compiler optimization assumption invalid: ERROR: failed to find exception throw blockFAILED: (add != nullptr)
    File: /coreclr/src/jit/codegencommon.cpp Line: 1852
    Image:

Aborted
```

**Comment by AndyAyersMS** [comment]
Failure case:
```
/home/andya/repos/coreclr/bin/tests/Linux.arm64.Checked/Tests/Core_Root/corerun /home/andya/repos/jitutils/bin/pmi.dll PREPONE-QUIET /home/andya/repos/coreclr/bin/tests/
Linux.arm64.Checked/Tests/Core_Root/System.Private.CoreLib.dll 14158
Prepone for System.Private.CoreLib method 14158
PREPONE type# 988 method# 14158 System.Runtime.Intrinsics.Arm.Arm64.Simd::Extract

Assert failure(PID 35728 [0x00008b90], Thread: 35728 [0x8b90]): Compiler optimization assumption invalid: ERROR: failed to find exception throw blockFAILED: (add != nullptr)
    File: /coreclr/src/jit/codegencommon.cpp Line: 1852
    Image:
```

**Comment by AndyAyersMS** [comment]
Asserts under `genHWIntrinsicSwitchTable` as there is no `SCK_ARG_RNG_EXCPN` block.

**Comment by AndyAyersMS** [comment]
The `BadImageFormat` exceptions above may be caused by my odd test setup, where I have some parts of the arm64 x8 change but not all. So will disregard for now.

**Comment by AndyAyersMS** [comment]
Aside from the above (which is also in #20260), there are 3 other assemblies with failures:

First is #20269
```
System.Reflection.Metadata

Assert failure(PID 42733 [0x0000a6ed], Thread: 42733 [0xa6ed]): Assertion failed '!"Incompatible types for gtNewTempAssign"' in 'System.Reflection.Metadata.Ecma335.SignatureDecoder`2[Vector`1,Int64][System.Numerics.Vector`1[System.Single],System.Int64]:DecodeType(byref,bool,int):struct:this' (IL size 388)
    File: /coreclr/src/jit/gentree.cpp Line: 14457
    Image: /home/andya/repos/coreclr/bin/tests/Linux.arm64.Checked/Tests/Core_Root/corerun
```
and the other two hit this assert in the runtime:
```
System.Security.Cryptography.OpenSsl
System.Security.Cryptography.Algorithms

Assert failure(PID 45202 [0x0000b092], Thread: 45202 [0xb092]): Consistency check failed: File has not had execution verifiedFAILED: m_bDisableActivationCheck || CheckLoadLevel(FILE_ACTIVE)
         FAILED: pDomainFile->CheckActivated()
                /coreclr/src/vm/ceeload.cpp, line: 13895
         FAILED: GetModule()->CheckActivated()
                /coreclr/src/vm/methodtable.cpp, line: 9346
         FAILED: CheckActivated()
                /coreclr/src/vm/methodtable.cpp, line: 3667
    File: /coreclr/src/vm/domainfile.cpp Line: 370

**Comment by echesakovMSFT** [comment]
The last two issues are https://github.com/dotnet/coreclr/issues/19668

**Comment by AndyAyersMS** [comment]
Seems like this should be running clean now. Let me investigate.

**Comment by AndyAyersMS** [comment]
Not quite. Still hit these cases (~4 methods in all, with overloads)
```
System.Runtime.Intrinsics.Arm.Arm64.Simd::Extract
System.Runtime.Intrinsics.Arm.Arm64.Simd::Insert

Assert failure(PID 8949 [0x000022f5], Thread: 8949 [0x22f5]): Compiler optimization assumption invalid: ERROR: failed to find exception throw blockFAILED: (add != nullptr)
    File: /coreclr/src/jit/codegencommon.cpp Line: 1856
```

**Comment by AndyAyersMS** [comment]
Other than that all the normal jit-diff fx assemblies run w/o assert on arm64, and all assemblies are assert free for jit-diffs crossgen.

**Comment by AndyAyersMS** [comment]
The remaining failures are HW intrinsic related.

The assert is triggrered by Arm64's `genHWIntrinsicSwitchTable` trying to range check a non-constant immediate for the `Extract` and `Insert` operators.

My impression is that any operation that wants to introduce check control flow late like this has to run an anticipatory step upstream, eg adding a bounds check node. But once you do that it will also expand to the right check, so the range check introduction in codegen is no longer needed.

So the idea is to add the equivalent of `addRangeCheckIfNeeded`  to the original intrinsic expansion, and then remove the codegen-added check.

@CarolEidt @tannergooding seem reasonable?

**Comment by AndyAyersMS** [comment]
@CarolEidt @tannergooding sorry to nag -- there is one HW intrinsic issue blocking a clean PMI of corelib on arm64. Suggested a fix above, let me know if this makes sense.

**Comment by tannergooding** [comment]
(sorry, missed the original ping)

That seems reasonable to me. We can always adjust it later if needed.

**Comment by tannergooding** [comment]
It might be worth noting that the suggested fix looks to be what we are already doing for the x86 intrinsics.
See `addRangeCheckIfNeeded`: https://github.com/dotnet/coreclr/blob/master/src/jit/hwintrinsicxarch.cpp#L651
and its usages:
* https://github.com/dotnet/coreclr/blob/master/src/jit/hwintrinsicxarch.cpp#L857
* https://github.com/dotnet/coreclr/blob/master/src/jit/hwintrinsicxarch.cpp#L873

**Comment by AndyAyersMS** [comment]
Right -- the idea was to do something similar for arm64.

**Comment by CarolEidt** [comment]
Sorry @AndyAyersMS I also missed this. I can take a look at it, but probably not until tomorrow.

**Comment by AndyAyersMS** [comment]
No worries, I have something coded up -- will have a PR out soon.
